### DISCUSSION TITLE
How to test Keycloak authentication locally on a mobile app or emulator without using localhost?

### DISCUSSION THREAD

**Discussion by develogo** [fa-fifi/keycloak_wrapper#45] [Q&A]
How can I test Keycloak authentication locally from a mobile app or emulator, considering that localhost is not accessible from the mobile device? For example, I'm using the following URL in my WebView:

`http://localhost:8080/realms/bettercity/protocol/openid-connect/auth?redirect_uri=com.example.demo%3A%2F%2Flogin-callback&client_id=go-api&response_type=code&prompt=login&state=KUrzCJR7OudVuKD6FVHhXQ&nonce=w2_ulXxVOAt3LAIZ1fiEqw&scope=openid%20profile%20offline_access&code_challenge=pEjJe4cED7WFXN9rKWopMXUblNOrj_Sqf4UyjvJxScE&code_challenge_method=S256`

What are the best practices for setting up the redirect URIs and configuring the server for local development testing, either through an emulator (using 10.0.2.2) or a physical device (using my local IP address)?

**Comment by fa-fifi** [comment]
I'm not sure what the best practices are since I never actually tested Keycloak authentication locally, but you can refer to [this](https://stackoverflow.com/questions/5528850/how-do-you-connect-localhost-in-the-android-emulator) to connect localhost in your emulator.

**Reply by develogo** [reply]
I've already done this, the problem is when opening the webview, even defining the local ip of another machine it points to "localhost" and not to the correct ip

**Comment by develogo** [comment]
https://github.com/user-attachments/assets/c3b7231e-65be-4eeb-a652-c44304d503cd

**Reply by fa-fifi** [reply]
Have you configured the port inside the emulator setting?
And add * inside all your redirect urls. → com.example.demo:/*

I'm not sure this will resolve your issue but let's try first.

**Reply by develogo** [reply]
My suspicion is that when using http it is redirecting to localhost

**Reply by fa-fifi** [reply]
Sorry for the late response. Is it resolved already?

**Comment by nouhouari** [comment]
Hi @develogo,
Don't use localhost but your IP address (192.168.0.10) or 10.0.2.2. 
It should work.
